Refactor ADT-like type classes (try 2)

Lysxia opened the pull request:
Hi,

I'm back with another bag of tricks.

- This time the user interface is not affected, as far as I could tell. The type signatures do not change at all, and I added a small test suite to check that the classification of `ADT` / `ADTRecord` / `ADT1` / `ADTRecord1` remains the same.
- `Generic*Profunctor` methods are separated into independent classes (but the original classes are kept as combinations of those). This makes orthogonal the constraints of number of constructors and number of fields of each constructor. This would not scale with the current implementation, since we would need one `ADT`-like type class for each combination of constraint. For example, it is possible to describe types which are purely sums of unary constructor, or constructors with at least one field. For the latter case, we can safely write a generic `foldMap` to a `Semigroup`. I have yet to find a good use for sums of unary constructors; a pure version of `consume` comes to mind, but you can implement that for a `ADTNonEmpty` by just forgetting extra fields.
- I managed to merge `ADT` and `ADT1`. It solves a bit of awkwardness around the handling of `K1` (which makes types like `(,)` not `ADTNonEmpty1` nor `ADTRecord1`), although some of it still remains at the moment for backwards compatibility. This again makes the implementation more expressive, allowing more possible one-liner combinators.

A bit of explanation on the actual implementation. We cannot compute the necessary profunctor constraints as a type-level **function** of the generic data type, otherwise we end up with the problem of my previous attempt where the interface leaks implementation details. So instead we give them upfront and take a **relational** approach.

Let `t` be a data type, and `ks :: [(* -> * -> *) -> Constraint]` be a list of profunctor type classes (`Profunctor`, `GenericProfunctor`...). There is an instance of `ADT_ Proxy ks (Rep t)` if, for any type `p :: * -> * -> *`, satisfying all the constraints in `ks` is sufficient to implement a traversal of the `Generic` type `t` (`(forall a. c a => p a a) -> p t t`). (Similarly with `ADT_ Identity ks (Rep1 t)` for `Generic1` types. Notice the instantiation of `unary` to `Proxy`/`Identity`.)

The `ks |- k` type class abstractly expresses that the conjunction of constraints `ks` entails the constraint `k` (if `p` satisfies `ks`, then `p` satisfies `k`), hence the turnstile-like notation, but concretely it's just the list membership relation because GHC doesn't have sufficient constraint-level reasoning for anything fancier. For that entailment relation to be complete, we assume that `ks` contains all the superclasses of its elements. Then, the superclass hierarchy of `Generic*Profunctor` becomes superfluous, hence the split I propose in this patch, which increases expressiveness at the same time.

An instance like

    instance (ks |- GenericProductProfunctor, ADT_ unary ks f, ADT_ unary ks g)
      => ADT_ unary ks (f :*: g) where

thus reads that if I can traverse `f` and `g` with some profunctor `p` satisfying `ks`, and if `p` satisfying `ks` implies that `p` is an instance of `GenericProductProfunctor` (providing `mult`), then I can traverse `f :*: g`.

sjoerdvisscher commented:
Awesome, you did it!

Some random thoughts while I try to comprehend all the details:

Would it work to merge the two `Constraints'` type families by having two class parameters? Then you might have:

```
type Constraints t c = Constraints' (Rep t) c AnyType
type Constraints1 t c = Constraints' (Rep1 t) Absurd c
```

Except that for `ADT1` you have `AnyType` instead of `Absurd`. When do you need `AnyType` and when `Absurd`?

Why did you call `generic1'` `nonEmpty1'`?

In the testfile, it should be possible to drop `Constraints1 t AnyType`.

What do you mean with "It solves a bit of awkwardness around the handling of K1 (which makes types like (,) not ADTNonEmpty1 nor ADTRecord1)" I had not noticed issues with `(,)`, what goes wrong currently? (I can't try it at the moment)
And "although some of it still remains at the moment for backwards compatibility.", what remains? Note that `one-liner` doesn't have to be backwards compatible, as long as the changes don't make it worse!

Lysxia commented:
> Would it work to merge the two Constraints' type families by having two class parameters?

Good idea. I merged them.

> In the testfile, it should be possible to drop Constraints1 t AnyType.

Indeed! Fixed.

> What do you mean with "It solves a bit of awkwardness around the handling of K1 (which makes types like (,) not ADTNonEmpty1 nor ADTRecord1)"

What I was referring to is that there are no instances of `ADTRecord1'` and `ADTNonEmpty1'` for `K1` ([source](https://github.com/sjoerdvisscher/one-liner/blob/34239f24fbd06eab6f036c252506374f8aa04cc7/src/Generics/OneLiner/Internal.hs#L115-L128)).

`K1` is used in `Rep1 ((,) a)`, or, for instance, in any record type with a monomorphic field. Thus `ADTRecord1'` and `ADTNonEmpty1'` are not applicable to these types.

Furthermore, in the current version, the instance of `ADT1'` for `K1` uses `identity`:

    instance ADT1' (K1 i v) where generic1' _ _ _ = dimap unK1 K1 identity

And one-liner makes good use of this to define `gmap1` and `gtraverse1`, specializing at `(->)` and `Star`.

However, by unifying `ADT1'` with `ADT'`, replacing `identity` in the `K1` instance with an explicit parameter to `generic_`, it is possible to define more general mappings, where one provides both `forall a. c a => a -> a` and `forall s a. c1 s => (a -> a) -> s a -> s a`.  
However, our current `ADT*1` functions don't have that parameter, so we need to give a default argument to `generic_`. In the case of `ADT1`, we have `identity`. In the case of `ADTNonEmpty1` and `ADTRecord1`, we need a dummy, which `Absurd` provides. These result in different implementations,  `generic1'` and `nonEmpty1'`. Initially, `generic1` wasn't present the patch, but I now added it for consistency).  
Just now, I also removed the `Absurd` type class, and instead rely on an additional parameter (`nullary`) to `ADT_` to indicate the availability of the `forall a. c a => p a a` argument, similarly to `unary`. (`nullary` and `unary` respectively refer to the arities of the arguments of `c` and `c1` above).
An alternative to all this is to modify the user facing functions to take an additional parameter, but I imagine that the current version can be useful too, which is a better motivation than the "backwards compatibility" I wrote earlier; we can always add the modified ones later.

Furthermore, from a more abstract perspective, this `identity :: GenericProfunctor p => p a a` being a common abstraction for `id` (`GenericProfunctor (->)`), `pure` (`GenericProfunctor (Star f)`) and `empty` (`GenericProfunctor (Joker f)`) is somewhat unsatisfactory.  
Note that for `Star f`, `const empty` is another potential implementation, so perhaps `identity` is representing two overlapping abstractions.  
If you compare `GenericProfunctor` to rings, there may be a correspondence between `(unit, identity)` and `(1, 0)` (there is also `zero`, but I consider it equivalent to `identity` from a certain point of view). This agrees with the `Joker` instance. However, `(->)` and `Star` break the following law: for all `x`, ``identity `mult` x = identity`` (`0 * x = 0`).  
This suggests that `id` (for `(->)`) and `pure` (for `Star`) belong elsewhere in a hierarchy of profunctors.   
The discrepancy can be made more obvious by generalizing both `identity` and `zero` to the following: `zero :: forall a a'. GenericProfunctor p => p a a'`.  
This may still be too abstract an issue, but I think it works out nicely with the unified instance of `ADT_` for `K1`, by removing a call to the problematic `identity`.

sjoerdvisscher commented:
> here are no instances of ADTRecord1' and ADTNonEmpty1' for K1

Hmm, I think that was an oversight on my part.
